torch/_functorch/aot_autograd.py

```diff
@@ -3357,6 +3655,264 @@ def forward(*runtime_args):
+    which is specified with `output_loss_index`.
+    All other outputs to the forward are presumed to not require gradients.
+(2) This API cannot capture optimizers (although in theory we could build an API for this).
+(3) Metadata mutations on params/buffers/inputs are banned.
+(4) Data mutations on anything that requires gradients are banned (parameters)
+(5) If an input is mutated, it is not allowed to alias any other inputs.
+(6) Parameters must not be duplicated.
+"""
+def aot_export_module(
+    mod: nn.Module,
+    args,
+    *,
+    decompositions: Optional[Dict] = None,
+    # If true, we'll return a joint forward-backward graph,
+    # As well as metadata on the loss + gradients in the backward.
+    trace_joint: bool,
+    # If trace_joint is True, we expect your module to return a scalar loss.
+    # Your module can return multiple outputs, so you must specify which output the loss is.
+    output_loss_index: Optional[int] = None,
+) -> Tuple[torch.fx.GraphModule, GraphSignature]:
+    named_parameters = dict(mod.named_parameters(remove_duplicate=False))
+    named_buffers = dict(mod.named_buffers(remove_duplicate=False))
+    params_and_buffers = {
+        **dict(named_parameters),
+        **dict(named_buffers),
+    }
+    params_and_buffers_flat, params_spec = pytree.tree_flatten(params_and_buffers)
+    params_and_buffers_flat = tuple(params_and_buffers_flat)
+    params_len = len(params_and_buffers_flat)
+
+    functional_call = create_functional_call(mod, params_spec, params_len)
+
+    num_fw_outs = None
+
+    if trace_joint:
+        # This helper effectively just adds some extra asserts about what the backward will look like:
+        # Outputs must include a scalar loss, that we compute gradients w.r.t.
+        # We don't compute gradients w.r.t. anything else: so just in case we detach()
+        # and other output tensors.
+        def fn_to_trace(*args):
+            nonlocal num_fw_outs
+            out = functional_call(*args)
+            if output_loss_index is not None:
+                assert isinstance(out, (list, tuple))
+                for i, o in enumerate(out):
+                    # We only want to create a backward graph w.r.t. the loss that the user passed in.
+                    # This implies that every other output should not require gradients.
+                    # Instead of making this an error (and forcing the user to detach all other outputs
+                    # of their forward),
+                    # we'll automatically detach them here.
+                    if o.requires_grad and i != output_loss_index:
+                        o.detach_()
+                out_loss = out[output_loss_index]
+                num_fw_outs = len(out)
+            else:
+                out_loss = out
+                num_fw_outs = 1
+            assert isinstance(out_loss, torch.Tensor)
+            assert out_loss.requires_grad
```

Comment:
Use of asserts here is inappropriate, because users can violate these asserts via inputs. Do real RuntimeErrors with proper error messages.